js/modules/form_input_textfield.js

```diff
@@ -1,122 +1,138 @@
+	 * @param bool passive
+	 */
+	markError(errorMessage, passive) {
+		window.log(this.properties.distinguisher, "Marking with error error message "+errorMessage, true);
 
-			this.pattern = this.element.getAttribute("pattern");
-			this.maxLength = this.element.getAttribute("maxlength");
-			this.disallowed = JSON.parse(this.element.getAttribute("data-disallowed"));
+		this.element.classList.add("invalid", "marked-invalid");
+		this.label.classList.add("active");
+		this.helperText.setAttribute("data-error", errorMessage);
 
-			// remove to prevent duplicates for BC instantiation methods
-			window.log(this.id, "Adding this.verify as a listener for input events if it was not already");
-			this.element.removeEventListener("input", this.verify.bind(this, true));
-			this.element.addEventListener("input", this.verify.bind(this, true));
+		if (!passive) {
+			M.escapeToast(errorMessage);
+			this.element.focus();
 		}
+	}
 
-		/**
-		 * @param errorType one of MISSING or INVALID
-		 * @param bool passive
-		 */
-		markError(errorType, passive) {
-			if (errorType == MISSING) {
-				var errorMessage = this.helperText.getAttribute("data-missing-error");
-				window.log(this.id, "Marking with error type MISSING, error message "+errorMessage, true);
-			} else if (errorType == INVALID) {
-				var errorMessage = this.helperText.getAttribute("data-invalid-error");
-				window.log(this.id, "Marking with error type INVALID, error message "+errorMessage, true);
-			} else {
-				throw "Invalid error type passed to "+className+".markError ("+errorType+")";
-			}
+	/**
+	 * @return string
+	 */
+	getValue() {
+		return this.element.value;
+	}
 
-			this.element.classList.add("invalid", "marked-invalid");
-			this.label.classList.add("active");
-			this.helperText.setAttribute("data-error", errorMessage);
+	/**
+	 * @return string
+	 */
+	getAggregationValue() {
+		return this.getValue();
+	}
 
-			if (!passive) {
-				M.escapeToast(errorMessage);
-				this.element.focus();
+	/**
+	 * @param bool passive If the form is actively verifying the content (and thus toasts/etc should show) or
+	 *	 false if verify is being called from input
+	 * @return bool
+	 */
+	verify(passive=false) {
+		let value = this.getValue();
+		window.log(this.properties.distinguisher, "Verifying with value "+JSON.stringify(value));
+
+		if (value.length) {
+			if (this.properties.maxlength && value.length > this.properties.maxlength) {
+				window.log(this.properties.distinguisher, "Value length "+value.length+" exceeds maximum length "+this.properties.maxlength, true);
+				this.markError(this.properties.invalidError, passive);
+				return false;
+			}
+			if (!(new RegExp(this.properties.pattern)).test(value)) {
+				window.log(this.properties.distinguisher, "Pattern "+this.properties.pattern+" does not match value", true);
+				this.markError(this.properties.invalidError, passive);
+				return false;
+			}
+			if (this.properties.disallowed.includes(value)) {
```

Comment:
Event if `properties` has a default value (possibly `{}`) in the constructor, this call will fail if `disallowed` comes back as `undefined`

Reply:
Properties is created by the class TextField, so they should be consistent.

Reply:
Fair enough